### ISSUE TITLE
fix path-segment cache buster to work when the incoming url doesn't have a busting token

### ISSUE THREAD

**Issue by mmerickel** [Pylons/pyramid#1584]
There's an issue with path-based busting right now where css may reference background images/fonts/etc, and these urls (inside your css) likely do not contain busting tokens unless you're using an asset pipeline to create them. This means that the current path-based busting (which is our default) will fail in a large number of cases.

```css
.logo {
    background: url('../images/logo.png');
}
```

We should fix/design path-busting to work when the requested URL does not contain a busting segment so that this image request will not fail similarly to how query-string busting does not fail if the query string doesn't contain the busting token.

I propose something like mangling the url to the point where we can tell if the token is in there or not, but I don't see how to do it right now in a reliable way without knowing anything about the token. We could get rid of the `PathSegmentCacheBuster` top-level api and only expose the `PathSegmentMd5CacheBuster` which can check for an md5 in the first subpath and only remove something looking like an md5.

Alternatively we just switch our default to query-string busting and remove path segment busting, letting people write their own using the very simple `ICacheBuster` api. I'm arguing that path-segment busting is primarily useful if you have a manifest of urls so that everything in your stack can agree on the url and/or whether a token is present.

ping @chrisrossi, @bertjwregeer, @dstufft

**Comment by dstufft** [comment]
You could check for the existence of a filename, so your ``match()`` method looks something like:

```python
def match(subpath):
    if os.path.exists(subpath):  # Actually do whatever we'd really need to check this
        return subpath
    else:
        # Do Some stuff to make subpath == to the un-bustered path
        return subpath
```

This won't be as performant (sadly) unless you use a lru cache or something on it so it doesn't have to do a file look up on every ``match()`` call, though hopefully those calls would be limited due to the long cache times, but it still might make sense to use the query based buster by default.

**Comment by mmerickel** [comment]
The information required to check for the path is not passed to `match` atm. It'd require the full pathspec with package prefix that generate receives. Other than that it does seem appropriate.

**Comment by mmerickel** [comment]
@chrisrossi pointed out in IRC that my example is not actually broken because the original url should contain the token, and this relative include would still be relative to a token.

For example, original url is `.../static/<token>/css/default.css`, and generated url is `.../static/<token>/images/logo.png`, with the caveat being that the token in the second url is not correct for `logo.png` for better or worse.

So that worries me less, but given that the tokens are not correct I'm still a little bothered by it.

- I guess if someone really cares then they should write their own `ICacheBuster`.
- Maybe we should still switch the default to query string to be more strictly correct here?

**Comment by chrisrossi** [comment]
If my memory serves me correctly, I initially wanted to make query string the default, but someone, I think it was @mmerickel, pointed out that some proxy caches ignore the query string.  ;)

**Comment by bertjwregeer** [comment]
CloudFlare ignores query strings by default unless you tell it not to IIRC.

**Comment by mmerickel** [comment]
I've recently implemented some cache busting in an external gulp pipeline and I've come to the conclusion that you cannot reliably do path-segment busting without an external pipeline like this. You'd need to start rewriting assets inside pyramid for it to be "correct" and ain't nobody got time for that.

For this reason I propose removing the default or changing it to the query string. We then make a `ManifestCacheBuster` which can load a standard manifest file or alternatively just make a cookbook entry about it.

The manifests I'm using right now are generated by `gulp-rev` and have a super simple format that pyramid could internalize.

```json
{
    "images/foo.png": "images/foo-1234567890.png"
}
```

**Comment by dstufft** [comment]
I'm also using ``gulp-rev``, I have https://github.com/pypa/warehouse/blob/master/warehouse/utils/static.py which does the manifest dance. If y'all want it I'm happy to contribute it (or not if you don't, or you want it done differently).

**Comment by mmerickel** [comment]
I actually have a full setup working without pyramid 1.6 using a pregenerator combined with a notfound view that attempts to do a subrequest on the non-cache-busted url. So it works whether or not the filenames on disk are mangled.

If we made a general purpose resolver we would need to handle that reverse-mapping in the buster intelligently to support people who dump mangled names to the static folder and those who do not dump mangled names. Just a heads up.

**Comment by mmerickel** [comment]
This has been solved by #2013 by adding a configurable `ManifestCacheBuster` and removing `PathSegmentCacheBuster`.
